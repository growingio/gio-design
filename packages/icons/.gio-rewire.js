module.exports = {
  projectName: 'gio-design-icons',
};
