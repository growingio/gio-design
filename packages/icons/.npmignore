src
svgs
templates
.svgrrc.js
