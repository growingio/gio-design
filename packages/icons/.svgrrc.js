module.exports = {
  dimensions: false,
  expandProps: 'end',
  semi: true,
  typescript: true,
  outDir: 'src',
  indexTemplate: (filePaths) => {
    const exportEntries = filePaths.map((filePath) => {
      const files = filePath.split('/');
      const basename = files[files.length - 1].split('.')[0];
      return `export { default as ${basename} } from './${basename}';`;
    });
    exportEntries.push("export { IconProps } from './interface';");
    return `${exportEntries.join('\n')}\n`;
  },
  jsx: {
    babelConfig: {
      plugins: [
        [
          '@svgr/babel-plugin-remove-jsx-attribute',
          {
            elements: ['svg'],
            attributes: ['className', 'id', 'fill'],
          },
        ],
        [
          '@svgr/babel-plugin-add-jsx-attribute',
          {
            elements: ['svg', 'path'],
            attributes: [
              {
                name: 'fill',
                value: 'currentColor',
              },
            ],
          },
        ],
      ],
    },
  },
};
