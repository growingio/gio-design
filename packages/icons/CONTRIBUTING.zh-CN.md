# 如何贡献

## 目录结构

```
├── CONTRIBUTING.zh-CN.md
├── LICENSE
├── README.md
├── README.zh-CN.md
├── dist
├── es
├── node_modules
├── package.json
├── src
├── svgs
└── templates
```

- `svgs` 存放设计师提供的 svg 文件；
- `src` 把 svg 文件转成的 React 代码，并存放在 `src` 目录下；
- `es` tsc 编译 src 中的文件，生成 ES Modules 输出到 es 目录。

## 添加（更新）图标过程

1. 拿到设计师提供的 svg 文件后，如果是填充的图标，文件名需要添加 `-filled` 后缀，并放在 `svgs` 目录下；
2. 运行 `yarn workspace @gio-design/icons generate`；
3. 运行 `yarn workspace @gio-design/icons build`；
4. 在 `website` 中预览效果。
