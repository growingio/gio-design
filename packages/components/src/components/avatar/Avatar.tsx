import React, { useState, useEffect, useRef } from 'react';
import classNames from 'classnames';
import { More } from '@gio-design/icons';
import { isNil } from 'lodash';
import Tooltip from '../tooltip';
import { AvatarProps } from './interfaces';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import composeRef from '../../utils/composeRef';

const Avatar = React.forwardRef<HTMLSpanElement, AvatarProps>((props: AvatarProps, ref: React.Ref<HTMLSpanElement>) => {
  const {
    className,
    size = 'default',
    droppable = false,
    children: userName,
    src,
    omit = true,
    displayTooltip = false,
    prefixCls: customizePrefixCls,
    placement = 'bottom',
    tooltipTitle,
    ...rest
  } = props;

  const [isImgExist, setIsImgExist] = useState<boolean>(src !== undefined);
  const [scale, setScale] = useState<number>(1);
  const nodeRef = useRef<HTMLSpanElement>(null);
  const childrenRef = useRef<HTMLSpanElement>(null);
  const mergedRef = composeRef(ref, nodeRef);

  useEffect(() => {
    if (nodeRef.current && childrenRef.current) {
      const nodeWidth = nodeRef.current.offsetWidth;
      const childrenWidth = childrenRef.current.offsetWidth;
      setScale(nodeWidth - 8 < childrenWidth ? (nodeWidth - 8) / childrenWidth : 1);
    }
  }, [userName]);

  const prefixCls = usePrefixCls('avatar', customizePrefixCls);
  const classString = classNames(className, prefixCls, {
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-df`]: size === 'default',
    [`${prefixCls}-lg`]: size === 'large',
    [`${prefixCls}-hg`]: size === 'huge',
  });

  const childrenStyle: React.CSSProperties = {
    transform: `scale(${scale})`,
  };

  const renderMore = () => {
    if (droppable) {
      return (
        <div className={classNames({ [`${prefixCls}-droppable`]: droppable })}>
          <More size="16px" color="#FFFFFF" />
        </div>
      );
    }
    return null;
  };

  const renderAvatar = () => {
    if (!!src && isImgExist) {
      return <img alt="avatar" src={src} onError={() => setIsImgExist(false)} />;
    }
    if (userName !== undefined && typeof userName === 'string') {
      const prefixUserName = omit && typeof userName === 'string' ? userName.trim()[0] : userName.trim();
      return (
        <span ref={childrenRef} style={childrenStyle}>
          {prefixUserName}
        </span>
      );
    }
    return null;
  };

  const renderTooltip = (child: React.ReactElement) =>
    displayTooltip ? (
      <Tooltip title={isNil(tooltipTitle) ? userName?.trim() : tooltipTitle} placement={placement}>
        {child}
      </Tooltip>
    ) : (
      child
    );

  return renderTooltip(
    // For Dropdown trigger will set Event on rest
    // eslint-disable-next-line react/jsx-props-no-spreading
    <span ref={mergedRef} className={classString} {...rest}>
      {renderMore()}
      {renderAvatar()}
    </span>
  );
});

export default Avatar;
