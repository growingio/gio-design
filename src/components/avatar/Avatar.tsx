import React, { useState, useEffect, useRef } from 'react';
import classNames from 'classnames';
import { MoreOutlined, UserOutlined } from '@gio-design/icons';
import { isNil, isUndefined } from 'lodash';
import Tooltip from '../tooltip';
import { AvatarProps } from './interfaces';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import composeRef from '../../utils/composeRef';

const Avatar = React.forwardRef<HTMLSpanElement, AvatarProps>((props: AvatarProps, ref: React.Ref<HTMLSpanElement>) => {
  const {
    className,
    size = 'default',
    droppable = false,
    children: userName,
    src,
    omit = true,
    displayTooltip = false,
    prefixCls: customizePrefixCls,
    placement = 'bottom',
    tooltipTitle,
    ...rest
  } = props;

  const [isImgExist, setIsImgExist] = useState<boolean>(src !== undefined);
  const [scale, setScale] = useState<number>(1);
  const nodeRef = useRef<HTMLSpanElement>(null);
  const childrenRef = useRef<HTMLSpanElement>(null);
  const mergedRef = composeRef(ref, nodeRef);

  useEffect(() => {
    setIsImgExist(!isUndefined(src));
  }, [src]);

  useEffect(() => {
    if (nodeRef.current && childrenRef.current) {
      const nodeWidth = nodeRef.current.offsetWidth;
      const childrenWidth = childrenRef.current.offsetWidth;
      setScale(nodeWidth - 8 < childrenWidth ? (nodeWidth - 8) / childrenWidth : 1);
    }
  }, [userName, isImgExist]);

  const prefixCls = usePrefixCls('avatar', customizePrefixCls);
  const classString = classNames(className, prefixCls, {
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-df`]: size === 'default',
    [`${prefixCls}-lg`]: size === 'large',
    [`${prefixCls}-hg`]: size === 'huge',
  });
  const iconSizeMap: { [K: string]: string } = {
    small: '10',
    default: '14',
    large: '24',
    huge: '35',
  };

  const childrenStyle: React.CSSProperties = {
    transform: `scale(${scale}) translateX(-50%)`,
  };

  const renderMore = () => {
    if (droppable) {
      return (
        <div className={classNames({ [`${prefixCls}-droppable`]: droppable })}>
          <MoreOutlined size="16px" color="#FFFFFF" />
        </div>
      );
    }
    return null;
  };

  const renderAvatar = () => {
    if (!!src && isImgExist) {
      return <img alt="avatar" src={src} onError={() => setIsImgExist(false)} />;
    }
    if (!userName) {
      return (
        <span className={`${prefixCls}-default`}>
          <UserOutlined color="#fff" size={iconSizeMap[size]} />
        </span>
      );
    }
    if (typeof userName === 'string') {
      const prefixUserName = omit && typeof userName === 'string' ? userName.trim()[0]?.toUpperCase() : userName.trim();
      return (
        <span ref={childrenRef} style={childrenStyle} className={`${prefixCls}-string`}>
          {prefixUserName}
        </span>
      );
    }
    return null;
  };

  const renderTooltip = (child: React.ReactElement) =>
    displayTooltip ? (
      <Tooltip title={isNil(tooltipTitle) ? userName?.trim() : tooltipTitle} placement={placement}>
        {child}
      </Tooltip>
    ) : (
      child
    );

  return renderTooltip(
    // For Dropdown trigger will set Event on rest
    // eslint-disable-next-line react/jsx-props-no-spreading
    <span ref={mergedRef} className={classString} {...rest}>
      {renderMore()}
      {renderAvatar()}
    </span>
  );
});

export default Avatar;
