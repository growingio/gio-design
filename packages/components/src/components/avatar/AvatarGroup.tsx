import React from 'react';
import classNames from 'classnames';
import _ from 'lodash';
import Avatar from './Avatar';
import { AvatarGroupProps, UserAvatarType } from './interfaces';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';

const AvatarGroup: React.FC<AvatarGroupProps> = (props: AvatarGroupProps) => {
  const { number = 5, users = [], className, placement = 'bottom', style, displayTooltip = true } = props;
  const prefixCls = usePrefixCls('avatar');

  let children = null;
  const renderAvatarGroup = (sliceUsers: UserAvatarType[]) =>
    sliceUsers.map((user) => (
      <Avatar key={user.name} displayTooltip={displayTooltip} placement={placement} {...user}>
        {user.name}
      </Avatar>
    ));
  const renderAvatarRest = (restUsers: UserAvatarType[]) => (
    <Avatar className={`${prefixCls}-rest`} omit={false}>{`+${restUsers.length}`}</Avatar>
  );
  const classString = classNames(className, `${prefixCls}-group`);

  if (users.length === 0) {
    return null;
  }
  if (users.length <= number) {
    children = renderAvatarGroup(users);
  } else {
    const sliceUsers = _.slice(users, 0, number - 1);
    const restUsers = _.slice(users, number - 1, users.length);
    children = (
      <>
        {renderAvatarGroup(sliceUsers)}
        {renderAvatarRest(restUsers)}
      </>
    );
  }

  return (
    <div className={classString} style={style}>
      {children}
    </div>
  );
};
export default AvatarGroup;
