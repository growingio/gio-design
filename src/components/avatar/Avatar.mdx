import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Avatar from './Avatar';

# Loading 加载中

<Subtitle>用来代表用户，支持图片或字符展示。</Subtitle>

## 参数说明
<ArgsTable of={Avatar} />

## 代码演示

### 基础头像

当用户未设置头像时，显示该用户名称的首个文字、数字或字母。

[Example](/?path=/story/functional-components-avatar--default)

### 尺寸

头像有以下几种尺寸：24px、32px、56px、80px。24px、32px 常用于列表&导航展示。56px、80px 常用于用户管理。默认为字符型头像，24px、32px 头像的文字字号为12px；56px 头像的文字字号为 16px；80px 头像的文字字号为 20px。

[Example](/?path=/story/functional-components-avatar--size)

### 可展开操作

hover头像可显示icon

[Example](/?path=/story/functional-components-avatar--hover)

### 重叠展示

可设置组显示数量，默认为4。当用户未设置头像时，显示该用户名称的首个文字数字或字母。hover头像时头像前置，并显示tooltip。

[Example](/?path=/story/functional-components-avatar--group)