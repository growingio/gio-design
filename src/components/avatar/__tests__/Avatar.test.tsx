import React from 'react';
import { render, screen } from '@testing-library/react';
import { PlusCircleFilled } from '@gio-design/icons';
// import renderer from 'react-test-renderer';
import Avatar from '../index';
import AvatarGroup from '../AvatarGroup';
import { Default } from '../Avatar.stories';

describe('Testing Avatar', () => {
  const users = [
    {
      name: 'li',
    },
    {
      name: 'pan',
    },
    {
      name: 'liu',
    },
    {
      name: 'wang',
    },
    {
      name: 'tong',
    },
    {
      name: 'leng',
    },
  ];

  it('Default', () => {
    render(<Default {...Default.args} />);
    expect(screen.queryByText('è¿™', { exact: false })).not.toBeNull();
  });

  it('props src', () => {
    const { container } = render(<Avatar>li</Avatar>);
    expect(container.getElementsByClassName('img')).toBeTruthy();
  });

  it('props size', () => {
    const { container } = render(<Avatar size="small" />);
    expect(container.getElementsByClassName('.gio-avatar-sm')).toBeTruthy();
  });

  it('props icon', () => {
    const { container } = render(<PlusCircleFilled />);
    expect(container.getElementsByClassName('.gio-avatar-icon')).toBeTruthy();
  });

  it('prop droppable', () => {
    const { container } = render(<Avatar size="small" droppable />);
    expect(container.getElementsByClassName('.gio-avatar-droppable')).toBeTruthy();
  });

  it('prop circle', () => {
    const { container } = render(<Avatar mode="circle" />);
    expect(container.getElementsByClassName('.gio-avatar-circle')).toBeTruthy();
  });

  it('not string', () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    // eslint-disable-next-line react/no-children-prop
    const { rerender } = render(<Avatar children={1} />);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    // eslint-disable-next-line react/no-children-prop
    rerender(<Avatar children={1} />);
    screen.logTestingPlaygroundURL();
    expect(screen).toBeTruthy();
  });

  it('props users', () => {
    const { container } = render(<AvatarGroup users={[]} number={4} placement="bottom" displayTooltip />);
    expect(container.getElementsByClassName('gio-avatar-group')).toBeTruthy();
  });

  it('should be stable', () => {
    const { rerender } = render(<AvatarGroup users={users} />);
    rerender(<AvatarGroup users={users} />);
  });

  it('should be render rightly', () => {
    const { container } = render(<AvatarGroup users={users} />);
    expect(container.getElementsByClassName('gio-avatar-group')).toBeTruthy();
  });

  it('props number', () => {
    const { container } = render(<AvatarGroup users={users} number={4} />);
    expect(container.getElementsByClassName('gio-avatar-rest')).toBeTruthy();
  });

  test('props placement', () => {
    const { rerender } = render(<AvatarGroup users={users} number={4} placement="top" />);
    rerender(<AvatarGroup users={users} number={4} placement="top" />);
  });

  it('props users', () => {
    const { container } = render(<AvatarGroup users={users.slice(0, 3)} number={4} />);
    expect(container.getElementsByClassName('gio-avata')).toBeTruthy();
  });
});
