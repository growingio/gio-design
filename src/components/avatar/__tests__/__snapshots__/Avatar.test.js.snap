// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Avatar should be stable 1`] = `
<span
  className="gio-avatar gio-avatar-df"
  style={
    Object {
      "backgroundColor": undefined,
    }
  }
>
  <span
    className="gio-avatar-string"
    style={
      Object {
        "transform": "scale(1) translateX(-50%)",
      }
    }
  >
    å¿«
  </span>
</span>
`;
